ffconv/stream_processors.py

```diff
@@ -0,0 +1,240 @@
+        """
+        # Set direct values from input and stream
+        self.input = input
+        self.index = stream['index']
+        self.codec = stream['codec_name']
+        self.language = stream.get('tags', {}).get('language')
+        if self.language == 'und':
+            self.language = None
+        self.allowed_codecs = profile[self.media_type]['codecs']
+
+        # Select target values from profile
+        self.target_codec = self.allowed_codecs[0]
+        self.target_container = profile[self.media_type]['container']
+        self.output = '{}-{}.{}'.format(self.media_type, self.index, self.target_container)
+
+        # Set stream-specific data
+        self._init_stream(input, stream, profile)
+
+    def _init_stream(self, input, stream, profile):
+        """
+        Set stream-specific input and target specs from input file, stream
+        and profile.
+
+        Must be overridden by subclasses.
+        """
+        return NotImplementedError('{} cannot set stream-specific data.'.format(self.__class__.__name__))
+
+    @property
+    def must_convert(self):
+        """
+        Default implementation for decision to convert the stream, based only
+        on codec compatibility.
+
+        Subclasses might modify or replace it completely.
+        """
+        return self.codec not in self.allowed_codecs
+
+    def clean_up(self):
+        """
+        Post-conversion stream clean-up, must be defined by subclasses.
+        """
+        raise NotImplementedError('{} cannot clean up {} yet.'.format(self.__class__.__name__, self.media_type))
+
+    def convert(self):
+        """
+        Conversion, must be defined by subclasses.
+        """
+        raise NotImplementedError('{} cannot convert {} yet.'.format(self.__class__.__name__, self.media_type))
+
+    def process(self):
+        """
+        Process this stream, which might mean converting it or not.
+
+        :return: processed stream data
+        """
+        log('Processing stream <index:{} type:{} codec:{}>...'.format(self.index, self.media_type, self.codec), 2)
+        if self.must_convert:
+            # Must convert, run conversion and clean-up
+            log('Converting...', 3)
+            self.convert()
+            log('Cleaning up file...', 3)
+            self.clean_up()
+            self.input = self.output
+            self.index = 0
+            log('Done', 2)
+
+        else:
+            # Nothing to do
+            log('Skipping, no conversion required', 3)
+
+        # Build and return data: input file, index and language (used by merger)
+        res = {'input': self.input, 'index': self.index}
+        if self.language:
+            res['language'] = self.language
+        return res
+
+
+class VideoProcessor(StreamProcessor):
+    """
+    Video stream processor, provides specific video conversion functionality.
+    """
+    media_type = 'video'
+
+    def _init_stream(self, input, stream, profile):
```

Comment:
![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png) Redefining built-in 'input' [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://localhost:9000/coding_rules#rule_key=Pylint%3AW0622)